### PR TITLE
refactor: remove `Expr.Def` in `Resolver` and later

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to https://github.com/flix/flix/pull/8641
Related to https://github.com/flix/flix/issues/8399

### CODE REVIEW

**Comment by jaschdoc** [comment]
!!!rerun

**Comment by jaschdoc** [comment]
@JonathanStarup @magnus-madsen would also love a review on this :)

File: main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala [hunk 1]
```diff
@@ -228,17 +228,17 @@ object EntryPoint {
...
     // NB: Getting the type directly from the scheme assumes the function is not polymorphic.
     // This is a valid assumption with the limitations we set on the entry point.
-    val func = TypedAst.Expr.Def(oldEntryPoint.sym, oldEntryPoint.spec.declaredScheme.base, SourceLocation.Unknown)
-
+    val itpe = oldEntryPoint.spec.declaredScheme.base
+    val defSym = Ast.DefSymUse(oldEntryPoint.sym, SourceLocation.Unknown)
+    val args = List(TypedAst.Expr.Cst(Constant.Unit, Type.Unit, SourceLocation.Unknown))
     // func()
-    val call = TypedAst.Expr.Apply(func, List(TypedAst.Expr.Cst(Constant.Unit, Type.Unit, SourceLocation.Unknown)), oldEntryPoint.spec.declaredScheme.base.arrowResultType, oldEntryPoint.spec.declaredScheme.base.arrowEffectType, SourceLocation.Unknown)
+    val call = TypedAst.Expr.ApplyDef(defSym, args, itpe, itpe.arrowResultType, itpe.arrowEffectType, SourceLocation.Unknown)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala:235]: defSym should be symUse here right?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala:235]: Just so we are on the same page- in what phase do we drop DefSymUse and just use DefnSym? Is it from SimplifiedAst and forward?

**Comment by magnus-madsen** [comment]
Conflicted. Sorry

**Comment by magnus-madsen** [comment]
I think this is almost good to go. I just want to ensure that we use "sym" and "symUse" correctly everywhere.

File: main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala [hunk 2]
```diff
@@ -101,16 +95,11 @@ object CodeHinter {
...
-          checkEffect(sym, lambda.tpe, loc)
-        case _ => Nil
-      }
-      hints0 ++ visitExp(exp) ++ visitExps(exps)
+      visitExp(exp) ++ visitExps(exps)
 
     case Expr.ApplyDef(Ast.DefSymUse(sym, loc1), exps, _, _, _, loc2) =>
       val hints0 = exps match {
-        case lambda :: _ => checkEffect(sym, lambda.tpe, loc2)
+        case lambda :: _ => checkEffect(sym, lambda.tpe, loc2) // why does this not check the rest of the exps?
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala:102]: I think it should!

File: main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala (hunk 1 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala:235]: Yes from `SimplifiedAst` and later it is just a `DefnSym`

**Comment by jaschdoc** [comment]
`symUse` should now be called `symuse` when it is a `DefSymUse` and otherwise a `sym` in the backend where it is a `Symbol.DefnSym`

File: main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala [hunk 3]
```diff
@@ -476,16 +476,9 @@ object Monomorpher {
...
-    case LoweredAst.Expr.Def(sym, tpe, loc) =>
-      /*
-       * !! This is where all the magic happens !!
-       */
-      val newSym = specializeDefSym(sym, subst(tpe))
-      MonoAst.Expr.Def(newSym, subst(tpe), loc)
-
     case LoweredAst.Expr.Sig(sym, tpe, loc) =>
       val newSym = specializeSigSym(sym, subst(tpe))
-      MonoAst.Expr.Def(newSym, subst(tpe), loc)
```
Comment (JonathanStarup) [inline main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala:488]: Matt is scared

**Comment by magnus-madsen** [comment]
@jaschdoc Conflicted, sorry.

**Comment by magnus-madsen** [comment]
@jaschdoc You accidentally resurrected MonoTypes which was removed :) 🧟 🧟 🧟

**Comment by jaschdoc** [comment]
Oops, I love git merges

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Thanks-- now we can progress on this track too :)

**Pull request merged**